doc/lpc18/testing.tex

```diff
@@ -55,25 +79,118 @@ \subsection{The Simple Test Framework}
+lists and individually inserted into the datapath.
+As the pcapng format, which contains port information as packet metadata, is 
+not yet supported by libpcap, we identify the input port of a pcap file by
+its filename.
+
+In addition to creating the runtime, the kernel version compiles the eBPF/XDP 
+program to an eBPF object using clang and llvm.\\
+\textbf{run}\\
+At this point, the user and kernel frameworks diverge in behavior.
+
+In the user space version, packets are converted to a mock-sk\_buff structure 
+and then directly passed into the to-be-tested eBPF/XDP function call. The 
+output result is recorded and written to per-port output pcap files.
+
+The kernel framework, however, is intended to verify the correctness of an 
+eBPF/XDP object in relation to the host kernel. This results in a more complex 
+pipeline (See Figure \ref{fig:kernel_test}). 
+Before the eBPF program is loaded, the framework creates a bridge running in an 
+network namespace. Isolation via a namespace ensures that existing 
+virtual interfaces and eBPF maps do not conflict with the test in question.
+\texttt{n} virtual interfaces are attached to the bridge, corresponding to 
+the highest port identifier defined in the STF file.
+
+The compiled eBPF object is then loaded into all ports using \texttt{tc} or 
+\texttt{ip}. Instead of directly calling into the eBPF program, the kernel 
+runtime writes the packet to the associated port using raw sockets. The output 
+results are recorded by sniffing on the interfaces via tcpdump.\\
+\textbf{check-results}\\
+After the program has completed, the results are collected and compared with 
+the expected packets defined by the STF template. If the output port or the 
+dataframe content does not match the specification, the test returns an error 
+code and is marked as failed.
+
+Both the user and kernel space runtime do not support testing for counters 
+and table state yet. We have planned these extensions as future work.
 \begin{figure}
 	\centering
 	\includegraphics[width=\linewidth]{user_test}
 	\caption{User-level testing of the C programs generated by the P4 compilers.}
 	\label{fig:user_test}
 \end{figure}
-
-\subsection{Testing P4 Programs End-to-End}
-Talk about how to use the test framework to verify your eBPF code or p4 file
-end-to-end. Highlight, how the framework can be used to test eBPF and XDP
-programs in general, independent from the P4 pipeline.
 \begin{figure}
 	\centering
 	\includegraphics[width=\linewidth]{kernel_test}
-	\caption{Kernel-level testing of the C programs generated by the P4 compilers.}
+	\caption{Kernel-level testing of the C programs generated by the P4 
+	compilers.}
 	\label{fig:kernel_test}
 \end{figure}
+\subsubsection{Using the Test Runtime}
+The test framework can be used independently to verify eBPF and XDP programs. 
+[TODO] Talk about how you test raw ebpf C programs with having a P4 template. 
```

Comment:
I don't get this TODO.

Reply:
That should have been a "without". The basic idea is that a person can also use our runtime to test their eBPF/XDP programs. They only need to provide their C code and stf file and we run it through the pipeline. This is not available yet, but should not be hard to implement. I think it is a nice use case. 
The other thing is there should be a way to instantly compile a P4 program to the eBPF byte code. 
Right now, you have to use `make -f p4c/backends/ebpf/runtime/kernel.mk BPFOBJ=out.o P4FILE=PROGRAM.p4` which is a bit awkward.
I think both of these features would also increase the usefulness of the compiler backend.